### PR TITLE
Implants, Surgery and Organs rework: Phase 1

### LINKED ISSUE #4298: You can have multiple surgical operations active on a patient.
I've done some self surgery tests and the "sheet" allowed me to start all the torso surgeries at the same time, which forced me to go through them all "guessing" every time which surgery was in progress.

Fix suggestions:
* Add a way to stitch the patient back together and cancel the surgery.
* Add a way to see which surgeries are in progress and select the relevant one or, do not allow more than one surgery at the same time.

### LINKED ISSUE #10212: Micro/Macrobombs just gib you without any explosion
:hamster:

### PR DESCRIPTION
**Pull request by Core0verload**
Implants:
- Implants, implant cases and implanters got RnD levels and can be DA'ed for science. 
- Implants and cases have RnD levels of implants they are containing.
- Implant stacking removed/changed. For example, loyality implants are not stackable, chem implants are stackable, freedom/adrenal/EMP implants are stacking their uses up to 2x from initial.
- Surgically removed implant can be placed into an implant case to be DAed or re-implanted. Remove implant with empty implant case in inactive hand.
- Fixes #10212, thanks @Anonus.
- #5781 is impossible now I think.
- Overall improvements in implant code.

Surgery:
- Surgical drape now shows only operations for selected body part.
- Surgical steps now got progress bars.
- You can cancel a surgery by using drape on operable body part again. This works only before you perform surgery's first step.
- You cannot start two surgeries on one body part at once.
- You can start two multiloc surgeries (i.e. two augumentations) on diffirent body parts at once.
- Operating computer shows next step for surgeries. No more alt-tabbing to view next step on wiki while operating.
- Fixes #4298.
- Overall improvements in surgery code. A lot of overall improvements in surgery code.

Organs:
- All internal organs are moved to type /obj/item/organ/internal. Except brains. For now.
- All organs have tech levels.
- Cyberimplants are organs now.
- Changeling eggs and xeno embryos are organs now.
- Abductor glands are organs now.
- Monkey <-> Human transformations now keep your organs.
- Changeling's "last resort" transfers ling's head organs to the monkey-chestburster. This would be used for ling's organ-based powers.
- New admin proc for editing mob's organs and implants.

**[This pull is ready to merge.]**

COMING SO0N (in Phase 2, Phase 3 or maybe never):
- Implanter, implant case and some implants in RnD. I'm thinking about putting freedom implant and adrenal implant to illegal 4 or 5.
- Merging of appendix removal, cyberimplantation and xeno removal into a single multizone organ manipulation surgery. Feel free to transplant abductor glands, xeno embryos or inflamed appendixes!
- A foldable self-surgery chair for traitor surgical kit.
- Eating organs. Cut out someone's heart and eat it raw!
- Some xenos abilities would be linked to their organs. Play with xenos' guts for fun and profit!
- Who knows?

### CODE REVIEW

**Comment by PKPenguin321** [comment]
sounds pretty cool, but im not sure what you mean by implant stacking being removed
does that mean you can't have more than a few implants at once? so you can't have an adrenal, freedom, and explosive implant at the same time?

**Comment by Supermichael777** [comment]
i think it means how many you can have in you?

**Comment by Rinoka** [comment]
remember kids: when in doubt, think about what you would want someone to be able to reasonably produce when given 3 of whatever you're giving to research

**Comment by Core0verload** [comment]
@Supermichael777, gang implants were self-destructing on injection. I didn't changed that. So they do not stay in body and you need to loot an unused implanter from a gangster to get the levels.

Uplink implant is not a common loot. It costs traitor all his TC, or all his TC and a lot of luck (implant kit). If you managed to get that implant, you deserve your insanely high illegal tech.

**Comment by Core0verload** [comment]
@PKPenguin321, it means you cannot run around with eight loyality implants, four uplink implants and two freedom implants in you. Second loyality implant will fail to inject, 4 uplink implants would be united into one uplink implant with 40TC in it, two freedom implants would be united into one implant with 2x uses.

File: code/game/objects/items/weapons/implants/implant_explosive.dm [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		imp_in.Weaken(7)
+	playsound(loc, 'sound/items/timer.ogg', 30, 0)
+	sleep(delay/4)
+	playsound(loc, 'sound/items/timer.ogg', 30, 0)
+	sleep(delay/4)
+	playsound(loc, 'sound/items/timer.ogg', 30, 0)
+	sleep(delay/4)
+	if(imp_in)
+		imp_in.gib()
+	explosion(src,heavy,medium,weak,weak, flame_range = weak)
```
Comment (Anonus) [inline code/game/objects/items/weapons/implants/implant_explosive.dm]: This needs to be above `if(imp_in)`

File: code/game/objects/items/weapons/implants/implant_explosive.dm [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	if(cause == "action_button" && alert(imp_in, "Are you sure you want to activate your microbomb implant? This will cause you to explode!", "Microbomb Implant Confirmation", "Yes", "No") != "Yes")
+		return 0
+	heavy = round(heavy)
+	medium = round(medium)
+	weak = round(weak)
+	imp_in << "<span class='notice'>You activate your microbomb implant.</span>"
+//If the delay is short, just blow up already jeez
+	if(delay <= 7)
+		imp_in.gib()
+		explosion(src,heavy,medium,weak,weak, flame_range = weak)
```
Comment (Anonus) [inline code/game/objects/items/weapons/implants/implant_explosive.dm]: This needs to be above `imp_in.gib()`

File: code/game/objects/items/weapons/implants/implant_explosive.dm (hunk 1 above)
Comment (Core0verload) [inline code/game/objects/items/weapons/implants/implant_explosive.dm]: It worked before, i just added if() check. Will change anyway.

**Comment by daman997** [comment]
>You can cancel a surgery by using drape on operable body part again. This works only before you perform surgery's first step.

thank you so much

**Comment by Anonus** [comment]
Okay, this fixes #10212 now.

**Comment by Core0verload** [comment]
And fixes #4298 as it adds a way to cancel surgeries. I'm still not sure about performing two surgeries on one body part at the same time. Must it be forbidden?

**Comment by PKPenguin321** [comment]
thanks for the explanation, that's much clearer
:+1:

**Comment by SconesC** [comment]
Wait, what about adrenal and freedom implants in R&D?

Other than that this looks fantastic

**Comment by Core0verload** [comment]
@SconesC, they are not in. I will or will not add them in Phase 2.

File: code/modules/surgery/helpers.dm [hunk 3]
```diff
@@ -80,13 +79,13 @@
...
 			face_covered |= I.flags_inv
-			eyesmouth_covered |= I.flags_cover
+			eyesmouth_covered |= I.flags
 		if(ishuman(C))
 			var/mob/living/carbon/human/H = C
 			for(var/obj/item/I in list(H.wear_suit, H.w_uniform, H.shoes, H.belt, H.gloves, H.glasses, H.ears))
 				covered_locations |= I.body_parts_covered
 				face_covered |= I.flags_inv
-				eyesmouth_covered |= I.flags_cover
+				eyesmouth_covered |= I.flags
```
Comment (phil235) [inline code/modules/surgery/helpers.dm]: MASKCOVERSMOUTH and other cover bitflags that we are going to check for are in flags_cover, not flags.

File: code/modules/mob/living/carbon/human/human.dm [hunk 4]
```diff
@@ -22,8 +22,8 @@
 					 /obj/item/organ/limb/r_arm, /obj/item/organ/limb/r_leg, /obj/item/organ/limb/l_leg)
 	for(var/obj/item/organ/limb/O in organs)
 		O.owner = src
-	internal_organs += new /obj/item/organ/appendix
-	internal_organs += new /obj/item/organ/heart
+	internal_organs += new /obj/item/organ/internal/appendix
+	internal_organs += new /obj/item/organ/internal/heart
 	internal_organs += new /obj/item/organ/brain
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/human.dm]: Why not changing the brain's path to organ/internal as well?

File: code/game/objects/items/weapons/implants/implant_explosive.dm [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+/obj/item/weapon/implant/explosive/get_data()
+	var/dat = {"<b>Implant Specifications:</b><BR>
+				<b>Name:</b> Robust Corp RX-78 Employee Management Implant<BR>
+				<b>Life:</b> Activates upon death.<BR>
+				<b>Important Notes:</b> Explodes<BR>
+				<HR>
+				<b>Implant Details:</b><BR>
+				<b>Function:</b> Contains a compact, electrically detonated explosive that detonates upon receiving a specially encoded signal or upon host death.<BR>
+				<b>Special Features:</b> Explodes<BR>
+				<b>Integrity:</b> Implant will occasionally be degraded by the body's immune system and thus will occasionally malfunction."}
```
Comment (phil235) [inline code/game/objects/items/weapons/implants/implant_explosive.dm]: The info about the integrity is incorrect, can you remove it?

File: code/game/objects/items/weapons/implants/implant_chem.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+				will suffer from an increased appetite.</B><BR>
+				<HR>
+				<b>Implant Details:</b><BR>
+				<b>Function:</b> Contains a small capsule that can contain various chemicals. Upon receiving a specially encoded signal<BR>
+				the implant releases the chemicals directly into the blood stream.<BR>
+				<b>Special Features:</b>
+				<i>Micro-Capsule</i>- Can be loaded with any sort of chemical agent via the common syringe and can hold 50 units.<BR>
+				Can only be loaded while still in its original case.<BR>
+				<b>Integrity:</b> Implant will last so long as the subject is alive. However, if the subject suffers from malnutrition,<BR>
+				the implant may become unstable and either pre-maturely inject the subject or simply break."}
```
Comment (phil235) [inline code/game/objects/items/weapons/implants/implant_chem.dm]: Same here, integrity info is outdated.

File: code/modules/mob/living/carbon/human/human.dm (hunk 4 above)
Comment (Core0verload) [inline code/modules/mob/living/carbon/human/human.dm]: Brains are special organs, at least until Phase 2.

**Comment by Core0verload** [comment]
It's already too big. Merge this before it starts conflicting please.

**Pull request merged**